Periodic improper support

kierannp opened the pull request:
<!-- Please confirm that your work is based on the correct branch. -->
<!-- Bug fixes should be based on *trunk-patch*. -->
<!-- Backwards compatible new features should be based on *trunk-minor*. -->
<!-- Incompatible API changes should be based on *trunk-major*. -->

## Description

Add support for a periodic improper force for 4 atoms bonded as an improper [https://manual.gromacs.org/current/reference-manual/functions/bonded-interactions.html#proper-dihedrals-periodic-type](url)

## Motivation and context

This is a common functional form in CHARMM-like forcefield and is an easy extension of the periodic dihedral already supported.

## How has this been tested?

Created python functions for calculating the force and energy resulting from this functional formal, see hoomd.md.pytest.test_improper.py for detailed test cases and functions to recreate test results. 


## Change log

<!-- Propose a change log entry. -->
```
Added test and support for periodic impropers

```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

kierannp committed:
Added notebook and basic improper tests talked about

kierannp committed:
Added copied improper force calculator files

kierannp committed:
Added Periodic improper class in python

kierannp committed:
Added dummy files

kierannp committed:
Added periodic improper test cases

kierannp committed:
Added python functions to generate test values

kierannp committed:
Added cpp periodic dihedral

CalCraven commented:
One thing to note about this PR is that the support for periodic impropers is important for running GAFF forcefields in HOOMD, which is recently supported in other engines via MoSDeF. One thing to note while @kierannp and I were working on this is the naming in HOOMD, which can be a bit confusing.
HOOMD has a periodic dihedral support already, which is called via hoomd.md.dihedral.Periodic. But the cpp code that is references is name [HarmonicDihedralForceCompute.cc(https://github.com/glotzerlab/hoomd-blue/blob/trunk-patch/hoomd/md/HarmonicDihedralForceCompute.cc). But again, it is a periodic dihedral force.

There already is a HarmonicImproperForceCompute.cc that computes a true a hoomd.md.improper.Harmonic functional form, and we are adding a hoomd.md.improper.Periodic. So we will call the cpp files PeriodicImproperForceCompute.cc. We may benefit at this time to rename the HarmonicDihedralForceCompute.cc (and like files) to PeriodicDihedralForceCompute.cc to reduce future confusion and potentially open the door for a true HarmonicDihedral force, which should look like any other harmonic force with an (a-a0)**2 type syntax, despite how [LAMMPS](https://docs.lammps.org/dihedral_harmonic.html) notates it with a cosine function, and uses the term `quadratic` to denote [harmonic type dihedrals](https://docs.lammps.org/dihedral_quadratic.html) to confuse all of us.

kierannp committed:
Periodic Improper test cases passing and compilation for CPU working

joaander commented on hoomd/md/PeriodicImproperForceCompute.h:
```diff
@@ -0,0 +1,108 @@
+#endif
+
+    protected:
+    Scalar* m_K;     //!< K parameter for multiple improper tyes
+    Scalar* m_sign;  //!< sign parameter for multiple improper types
+    int* m_multi;    //!< multiplicity parameter for multiple improper types
+    Scalar* m_chi_0; //!< chi_0 parameter for multiple improper types
```
Replace this with a GPUArray of `periodic_improper_params`, similar to bond parameters in https://github.com/glotzerlab/hoomd-blue/blob/0048a57dec73de737f0f53a4a480dd31ef19659e/hoomd/md/PotentialBond.h#L64

Doing this will reduce duplicate code in the later GPU implementation.

joaander commented on hoomd/md/improper.py:
```diff
@@ -110,3 +110,46 @@ def __init__(self):
+            * ``d`` (`float`, **required**), sign factor
+              :math:`\\n` :math:`[\\mathrm{dimensionless}]`.
+
+    Example::
+
+        harmonic = hoomd.md.improper.Periodic()
+        harmonic.params['A-B-C-D'] = dict(k=1.0, n = 1, phi0=0)
```
Make these changes to test the example code snippet using [sybil](https://sybil.readthedocs.io).

```suggestion
    .. rubric:: Example:

    .. code-block:: python

        periodic = hoomd.md.improper.Periodic()
        periodic.params['A-B-C-D'] = dict(k=1.0, n = 1, phi0=0)
        
        simulation.integrator.forces = [periodic]
```
and set the parameters appropriately.

Also, add the following to the top of the file in the module level docstring:
```
.. skip: next if(not hoomd.version.md_built)

.. invisible-code-block: python

    if hoomd.version.md_built:
        simulation = hoomd.util.make_example_simulation()
        simulation.operations.integrator = hoomd.md.Integrator(dt=0.001)
```

kierannp replied:
> Make these changes to test the example code snippet using [sybil](https://sybil.readthedocs.io).
> 
> and set the parameters appropriately.
> 
> Also, add the following to the top of the file in the module level docstring:
> 
> ```
> .. skip: next if(not hoomd.version.md_built)
> 
> .. invisible-code-block: python
> 
>     if hoomd.version.md_built:
>         simulation = hoomd.util.make_example_simulation()
>         simulation.operations.integrator = hoomd.md.Integrator(dt=0.001)
> ```

This example code produces the error:

```AttributeError: 'Simulation' object has no attribute 'integrator' ```

Do you have a suggestion for a different example case?

joaander replied:
You can find many examples scattered throughout the HOOMD documentation, search for `.. invisible-code-block: python`. I am certain this code works. Here is a test:
```
>>> simulation = hoomd.util.make_example_simulation()
>>> simulation.operations.integrator = hoomd.md.Integrator(dt=0.001)
>>>
```

joaander commented on hoomd/md/PeriodicImproperForceCompute.cc:
```diff
@@ -0,0 +1,373 @@
+        assert(improper.tag[3] <= m_pdata->getMaximumTag());
+
+        // transform a, b, and c into indices into the particle data arrays
+        // MEM TRANSFER: 6 ints
```
```suggestion
```

CalCraven replied:
Reminder to remove this line.

joaander commented on hoomd/md/improper.py:
```diff
@@ -110,3 +110,46 @@ def __init__(self):
                 len_keys=1))
         self._add_typeparam(params)
+
+class Periodic(Improper):
```
Add this class name to the list in `module-md-improper.rst` so the documentation appears.

joaander commented on hoomd/md/pytest/test_improper.py:
```diff
@@ -7,8 +7,56 @@
 
 import itertools
+
+# these series of functions are used to calculate the values for the test cases used below
```
These don't seem to be used below. I would suggest either using them in this method or preparing a separate `generate_periodic_improper_test_cases.py` file that uses them to generate the test cases used below.

CalCraven replied:
@kierannp I think putting these into a seperate generate_periodic_improper_test_cases.py make sense, and then we can import them to generate the force and energy terms used in the test cases. We could then easily add a test case for a few different positions, as well as future users would be able to use the dchi_dr functions to test forces for any future functional forms using chi. Also tagging @calvinpozderac who actually got these derivatives using mathematica.

joaander replied:
Now that they are called in the test, I'm fine with leaving them here. I would suggest wrapping them as static methods in a class to put them in a namespace that won't be confused with other tests:

```
class PeriodicImproperMath:
   @staticmethod
   def dchi_dr1(...)
```

kierannp replied:
I added them to a separate file as suggested. I think this format is more clean.

joaander reviewed (commented):
This looks like a good start! Here are some suggestions for improvement.

joaander commented:
> There already is a HarmonicImproperForceCompute.cc that computes a true a hoomd.md.improper.Harmonic functional form, and we are adding a hoomd.md.improper.Periodic. So we will call the cpp files PeriodicImproperForceCompute.cc. We may benefit at this time to rename the HarmonicDihedralForceCompute.cc (and like files) to PeriodicDihedralForceCompute.cc to reduce future confusion and potentially open the door for a true HarmonicDihedral force, which should look like any other harmonic force with an (a-a0)**2 type syntax, despite how [LAMMPS](https://docs.lammps.org/dihedral_harmonic.html) notates it with a cosine function, and uses the term `quadratic` to denote [harmonic type dihedrals](https://docs.lammps.org/dihedral_quadratic.html) to confuse all of us.

Yes, feel free to rename `HarmonicDihedralForceCompute*` to `PeriodicDihedralForceCompute*` at the C++ level, both with the file and class names. Please make this change in a separate pull request.

kierannp committed:
Added suggested changes except GPUArray due to issues with memory allocation

kierannp committed:
Corrected formatting

kierannp committed:
correct the corrected formatting

CalCraven reviewed (changes requested):
Looking good so far.

kierannp committed:
Finished modifications suggest in PR review

kierannp committed:
Improper documentation

kierannp committed:
Separate file for improper test case generation

kierannp committed:
Corrected formatting and documentation

kierannp committed:
Put generate function back into test file

CalCraven reviewed (changes requested):
Looking good, I think there are two more of Josh's edits still to get at. Also add your name to [this](https://github.com/glotzerlab/hoomd-blue/blob/0048a57dec73de737f0f53a4a480dd31ef19659e/sphinx-doc/credits.rst) file as well, should we add a note to the changelog?

joaander commented:
> should we add a note to the changelog?

Please don't. HOOMD has so many PRs that all the change log entries would conflict. I build the change log at release time from the drafts in each of the PR descriptions.

kierannp committed:
Created wrapper class for periodic improper math; add GPU support for periodic improper; made modification for suggestions; altered test case in documentation to not produce error

kierannp committed:
Corrected harmonic to periodic

kierannp committed:
Changed documentation example to match harmonic example otherwise it throws error/warning

kierannp committed:
Added names to list of contributers

tcmoore3 commented on hoomd/md/PeriodicImproperForceGPU.cu:
```diff
@@ -0,0 +1,380 @@
+        Scalar dthy = gbb * bby;
+        Scalar dthz = gbb * bbz;
+
+        // Scalar df = -K * dfab;
```
```suggestion
```

joaander committed:
Update hoomd/md/PeriodicImproperForceGPU.cu

Co-authored-by: Tim Moore <mtimc@umich.edu>

joaander committed:
Use periodic_improper_params in the base and GPU classes.

Also complete misc cleanup to bring the code up to modern HOOMD
coding standards.

joaander reviewed (approved):
Thanks for the complete PR and extensive tests.

I pushed a commit that moves the parameters into a GPUArray to reduce code duplication and the chances for errors in that duplication. I also made a few other changes to bring error messages up to modern HOOMD coding standards.

Could you double check my changes and let me know if they are correct? I got a little lost with the parameters. It seems that "d" and "sign" are the same parameter. It also seems that "n", "multi", and "m" are also all the same. You might consider modifying the code to keep the same names throughout to avoid confusion.

joaander committed:
Run clang-format.

kierannp commented:
> Thanks for the complete PR and extensive tests.
> 
> I pushed a commit that moves the parameters into a GPUArray to reduce code duplication and the chances for errors in that duplication. I also made a few other changes to bring error messages up to modern HOOMD coding standards.
> 
> Could you double check my changes and let me know if they are correct? I got a little lost with the parameters. It seems that "d" and "sign" are the same parameter. It also seems that "n", "multi", and "m" are also all the same. You might consider modifying the code to keep the same names throughout to avoid confusion.

The changes you made appear correct. I tried to use the same naming conventions as in similar scripts in HOOMD for calculating forces. I will look into making these variables names more consistent and explicit.

CalCraven commented:
> > It seems that "d" and "sign" are the same parameter. It also seems that "n", "multi", and "m" are also all the same. 
> 
> I tried to use the same naming conventions as in similar scripts in HOOMD for calculating forces. 

To clarify a bit more, d is named sign because d is always +-1. n is the multiplicity of the cos function. I would second the fact that we should stick to one value within hoomd. I believe these were derived from [this line](https://github.com/glotzerlab/hoomd-blue/blob/3179f17d71b85aad8024a66f6cd55853feb86fa2/hoomd/md/HarmonicDihedralForceCompute.cc#L252C45-L252C45). When I pull a PR to rename the HarmonicDihedralForce to PeriodicDihedralForce, we can make the same consistent name changes to those files as well.

kierannp committed:
changed variable names to single format
